### ISSUE TITLE
Code Python pour connaitre poids et biais

### ISSUE THREAD

**Issue by angv5845** [angv5845/Projet_IA_calcul_embarque#11]
import re
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.callbacks import EarlyStopping

def parse_c_array(filename, name, shape):
    """Lit un array C 'const float name[...] = { … };' et reshape en numpy."""
    text = open(filename, "r").read()
    m = re.search(rf'{re.escape(name)}\s*\[.*?\]\s*=\s*\{{(.+?)\}}\s*;', text, re.S)
    if not m:
        raise ValueError(f"{name} non trouvé dans {filename}")
    vals = re.findall(r'[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?\d+)?', m.group(1))
    arr = np.array([float(v) for v in vals], dtype=np.float32)
    exp = np.prod(shape)
    if arr.size < exp:
        print(f"⚠️ trouvé {arr.size} valeurs pour '{name}', padding 0.")
        arr = np.pad(arr, (0, exp - arr.size), 'constant')
    elif arr.size > exp:
        print(f"⚠️ trouvé {arr.size} valeurs pour '{name}', tronqué.")
        arr = arr[:exp]
    return arr.reshape(shape)

def build_model():
    inp = keras.Input(shape=(48,13,1))
    x = layers.Conv2D(8, 3, padding="same", activation="relu", name="conv2d")(inp)
    x = layers.MaxPool2D(2)(x)
    x = layers.Conv2D(16, 3, padding="same", activation="relu", name="conv2d_1")(x)
    x = layers.MaxPool2D(2)(x)
    x = layers.Flatten()(x)
    x = layers.Dense(32, activation="relu", name="bottleneck")(x)
    x = layers.Dense(8,  activation="relu", name="dense1")(x)
    out = layers.Dense(1, activation="sigmoid", name="dense2")(x)
    model = keras.Model(inp, out)
    model.compile(optimizer="adam", loss="mse", metrics=["mse"])
    return model

def export_layer_weights(layer, prefix, cfile, all_w, all_b):
    w = layer.get_weights()
    if not w:
        return
    kern, bias = w
    fk = kern.flatten()
    fb = bias.flatten()
    # On stocke dans les listes globales
    all_w.extend(fk.tolist())
    all_b.extend(fb.tolist())
    # On écrit le bloc C/PROGMEM pour cette couche
    cfile.write(f"// --- {prefix} weights & bias ---\n")
    cfile.write(f"const float {prefix}_w[] PROGMEM = {{\n")
    for v in fk:
        cfile.write(f"  {v:+.6f}f,\n")
    cfile.write("};\n\n")
    cfile.write(f"const float {prefix}_b[] PROGMEM = {{\n")
    for v in fb:
        cfile.write(f"  {v:+.6f}f,\n")
    cfile.write("};\n\n")

def main():
    # 1) Chargement
    X_train = parse_c_array("training.txt", "inputs",   (100,48,13))
    y_train = np.array([0]*50 + [1]*50, dtype=np.float32)
    X_test  = parse_c_array("test.txt",    "testData",(10,48,13))
    y_test  = np.array([0]*5  + [1]*5,   dtype=np.float32)

    # reshape pour Keras
    X_train = X_train[..., np.newaxis]
    X_test  = X_test[...,  np.newaxis]

    # 2) Construction & entraînement
    model = build_model()
    es = EarlyStopping(monitor="val_mse", patience=10, restore_best_weights=True)
    model.fit(
        X_train, y_train,
        validation_data=(X_test, y_test),
        epochs=200, batch_size=16,
        callbacks=[es], verbose=2
    )

    # 3) Évaluation générale
    loss, mse = model.evaluate(X_test, y_test, verbose=0)
    print(f"\nMSE test = {mse:.6f}")
    print("✅ ET7 validé !" if mse < 0.05 else "❌ À améliorer…")

    # 4) Vérification échantillon par échantillon
    print("\n--- Classification testdata détaillée ---")
    preds = model.predict(X_test)
    for i, p in enumerate(preds[:,0]):
        label = "Bleu" if p >= 0.5 else "Rouge"
        print(f"idx={i:>2} pred={p:.6f} → {label:5s} (attendu={y_test[i]})")

    # 5) Export des poids/biais + concaténés
    all_weights, all_biases = [], []
    with open("ET7.txt","w") as cfile:
        for layer in model.layers:
            export_layer_weights(layer, layer.name, cfile, all_weights, all_biases)
        # concaténation globale
        cfile.write("// --- ALL_WEIGHTS concat ---\n")
        cfile.write("const float all_weights[] PROGMEM = {\n")
        for v in all_weights:
            cfile.write(f"  {v:+.6f}f,\n")
        cfile.write("};\n\n")
        cfile.write("// --- ALL_BIASES concat ---\n")
        cfile.write("const float all_biases[] PROGMEM = {\n")
        for v in all_biases:
            cfile.write(f"  {v:+.6f}f,\n")
        cfile.write("};\n\n")

if __name__ == "__main__":
    main()
